File: apps/api/app/api/v1/tags.py
"""Tags API endpoints."""

from __future__ import annotations

from typing import TYPE_CHECKING
from uuid import UUID

from fastapi import APIRouter, Depends, status

from app.api.deps import get_current_user, get_db
from app.db.models import User
from app.modules.tags.schemas import TagCreate, TagResponse, TagUpdate
from app.modules.tags.service import TagService

if TYPE_CHECKING:
    from sqlalchemy.ext.asyncio import AsyncSession

router = APIRouter()


@router.post(
    "",
    response_model=TagResponse,
    status_code=status.HTTP_201_CREATED,
)
async def create_tag(
    data: TagCreate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> TagResponse:
    """Create a new tag for the authenticated user."""
    return await TagService.create_tag(db, current_user.id, data)


@router.get(
    "",
    response_model=list[TagResponse],
)
async def list_tags(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> list[TagResponse]:
    """Retrieve all tags belonging to the authenticated user."""
    return await TagService.list_tags(db, current_user.id)


@router.get(
    "/{id}",
    response_model=TagResponse,
)
async def get_tag(
    id: UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> TagResponse:
    """Retrieve details of a specific tag by ID."""
    return await TagService.get_tag(db, current_user.id, id)


@router.patch(
    "/{id}",
    response_model=TagResponse,
)
async def update_tag(
    id: UUID,
    data: TagUpdate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> TagResponse:
    """Update metadata of an existing tag."""
    return await TagService.update_tag(db, current_user.id, id, data)


@router.delete(
    "/{id}",
    status_code=status.HTTP_204_NO_CONTENT,
)
async def delete_tag(
    id: UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> None:
    """Delete a tag belonging to the authenticated user."""
    await TagService.delete_tag(db, current_user.id, id)

File: apps/web/lib/api/tags.ts
import axios, { AxiosInstance } from "axios";

const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || "http://localhost:8000";
const api: AxiosInstance = axios.create({
  baseURL: API_BASE_URL,
  withCredentials: true,
});

// ============ Types ============

export interface TagResponse {
  id: string;
  name: string;
  color: string;
  entry_count: number;
  created_at: string;
}

export interface TagCreate {
  name: string;
  color?: string;
}

export interface TagUpdate {
  name?: string;
  color?: string;
}

// ============ Tags API ============

export const tagsApi = {
  /**
   * Retrieve all tags belonging to the authenticated user
   */
  list: async (): Promise<TagResponse[]> => {
    const response = await api.get("/api/v1/tags");
    return response.data;
  },

  /**
   * Create a new tag for the authenticated user
   */
  create: async (data: TagCreate): Promise<TagResponse> => {
    const response = await api.post("/api/v1/tags", data);
    return response.data;
  },

  /**
   * Retrieve details of a specific tag by ID
   */
  get: async (id: string): Promise<TagResponse> => {
    const response = await api.get(`/api/v1/tags/${id}`);
    return response.data;
  },

  /**
   * Update metadata of an existing tag
   */
  update: async (id: string, data: TagUpdate): Promise<TagResponse> => {
    const response = await api.patch(`/api/v1/tags/${id}`, data);
    return response.data;
  },

  /**
   * Delete a tag
   */
  delete: async (id: string): Promise<void> => {
    await api.delete(`/api/v1/tags/${id}`);
  },
};
